src/library/scala/collection/immutable/ListSet.scala

```diff
@@ -174,9 +170,17 @@ sealed class ListSet[A] extends AbstractSet[A]
 
     /** `-` can be used to remove a single element from a set.
      */
-    override def -(e: A): ListSet[A] = if (e == head) self else {
-      val tail = self - e; new tail.Node(head)
-    }
+    override def -(e: A): ListSet[A] = removeInternal(e, this, Nil)
+
+    @tailrec private def removeInternal(k: A, cur: ListSet[A], acc: List[ListSet[A]]): ListSet[A] =
+      if (cur.isEmpty)
+        acc.last
+      else if (k == cur.head)
+        (cur.tail /: acc) {
```

Comment:
To be honest, it uses `/:` because I just copy-pasted and adjusted the `ListMap` code. If I was to write the code from scratch I'd probably use `foldLeft` (I also find `/:` a bit confusing) or even an imperative-style loop for performance reasons. As I also agree that standard library code is expected to be read by people with some expertise in the language (compared to other languages, Scala is actually very beginner-friendly on that matter!), I'm perfectly ok with both leaving this as it is now or rewriting it with a different style.